agent/consul/state/catalog.go

```diff
@@ -2546,125 +2546,109 @@ func (s *Store) updateTerminatingGatewayNamespace(tx *memdb.Txn, gateway structs
-	wc, err := s.serviceTerminatingGateway(tx, structs.WildcardSpecifier, entMeta)
-	if err != nil {
-		return fmt.Errorf("gateway service lookup failed: %s", err)
-	}
-	if wc != nil {
-		cfg := wc.(*structs.GatewayService)
-		mapping.CAFile = cfg.CAFile
-		mapping.CertFile = cfg.CertFile
-		mapping.KeyFile = cfg.KeyFile
-	}
-
+func (s *Store) updateGatewayService(tx *memdb.Txn, idx uint64, mapping *structs.GatewayService) error {
 	// Check if mapping already exists in table if it's already in the table
 	// Avoid insert if nothing changed
-	existing, err := s.serviceTerminatingGateway(tx, service, entMeta)
+	existing, err := tx.First(gatewayServicesTableName, "id", mapping.Gateway, mapping.Service)
 	if err != nil {
 		return fmt.Errorf("gateway service lookup failed: %s", err)
 	}
 	if gs, ok := existing.(*structs.GatewayService); ok && gs != nil {
+		mapping.CreateIndex = gs.CreateIndex
 		if gs.IsSame(mapping) {
 			return nil
 		}
+	} else {
+		// We have a new mapping
+		mapping.CreateIndex = idx
 	}
+	mapping.ModifyIndex = idx
 
-	if err := tx.Insert(terminatingGatewayServicesTableName, mapping); err != nil {
+	if err := tx.Insert(gatewayServicesTableName, mapping); err != nil {
 		return fmt.Errorf("failed inserting gateway service mapping: %s", err)
 	}
 
-	if err := indexUpdateMaxTxn(tx, idx, terminatingGatewayServicesTableName); err != nil {
-		return fmt.Errorf("failed updating terminating-gateway-services index: %v", err)
+	if err := indexUpdateMaxTxn(tx, idx, gatewayServicesTableName); err != nil {
+		return fmt.Errorf("failed updating gateway-services index: %v", err)
 	}
 	return nil
 }
 
-func (s *Store) serviceTerminatingGateway(tx *memdb.Txn, name string, entMeta *structs.EnterpriseMeta) (interface{}, error) {
-	return tx.First(terminatingGatewayServicesTableName, "service", structs.NewServiceID(name, entMeta))
+// serviceGateways returns all GatewayService entries with the given service name. This effectively looks up
+// all the gateways mapped to this service.
+func (s *Store) serviceGateways(tx *memdb.Txn, name string, entMeta *structs.EnterpriseMeta) (memdb.ResultIterator, error) {
+	return tx.Get(gatewayServicesTableName, "service", structs.NewServiceID(name, entMeta))
 }
 
-func (s *Store) terminatingGatewayServices(tx *memdb.Txn, name string, entMeta *structs.EnterpriseMeta) (memdb.ResultIterator, error) {
-	return tx.Get(terminatingGatewayServicesTableName, "gateway", structs.NewServiceID(name, entMeta))
+func (s *Store) gatewayServices(tx *memdb.Txn, name string, entMeta *structs.EnterpriseMeta) (memdb.ResultIterator, error) {
+	return tx.Get(gatewayServicesTableName, "gateway", structs.NewServiceID(name, entMeta))
 }
 
-func (s *Store) serviceTerminatingGatewayNodes(tx *memdb.Txn, service string, entMeta *structs.EnterpriseMeta) (structs.ServiceNodes, <-chan struct{}, error) {
+// TODO(ingress): How to handle index rolling back when a config entry is
```

Comment:
How come you added an index for these entries? Where would we use it?

Reply:
We use this index in `CheckIngressServiceNodes` to determine which has the bigger relevant index, `gateway-services` or `services` table and make sure we take the gateway-services table into account on blocking queries.

Consider the following scenario:
1. A state store with the following entries:
```
gateway-services:
- Service: foo, Gateway: ingress1, Index: 1
- Service: bar, Gateway: ingress2, Index: 2

services:
- Service: foo, Index: 3
- Service: ingress, Index: 4
- Service: bar, Index: 5
- Service: ingress2, Index: 6
```

2. A initial call to `CheckIngressServiceNodes(foo)` returns `Service: ingress, Index: 4`
3. A blocking query on `CheckIngressServiceNodes(foo)` is initiated with `MinIndex == 4`
4. The config entry of `ingress2` is updated to include `foo`, resulting in a new state store:
```
gateway-services:
- Service: foo, Gateway: ingress1, Index: 1
- Service: bar, Gateway: ingress2, Index: 2
- Service: foo, Gateway: ingress2, Index: 7

services:
- Service: foo, Index: 3
- Service: ingress, Index: 4
- Service: bar, Index: 5
- Service: ingress2, Index: 6
```

5. We would now expect to unblock the query and send an update of:
```
Index: 7
- Service: ingress
- Service: ingress2
```
which we can only do by looking at the indexes of the relevant gateway-services.